OpenRA.Mods.Common/Graphics/ContrailRenderable.cs

```diff
@@ -73,11 +73,31 @@ public void Render(WorldRenderer wr)
 			// Start of the first line segment is the tail of the list - don't smooth it.
 			var curPos = trail[Index(next - skip - 1)];
 			var curColor = color;
-			for (var i = 0; i < length - skip - 4; i++)
+
+			for (var i = 1; i < renderLength; i++)
 			{
-				var j = next - skip - i - 2;
-				var nextPos = Average(trail[Index(j)], trail[Index(j - 1)], trail[Index(j - 2)], trail[Index(j - 3)]);
-				var nextColor = Exts.ColorLerp(i * 1f / (length - 4), color, Color.Transparent);
+				var j = next - skip - 1 - i;
+				WPos nextPos;
+				int maxSmoothLength = 4; // Use 4 contrail position in circular buffer for max smooth effect.
+
+				var alpha = (renderLength - i) * color.A / renderLength;
+				if (alpha == 0)
```

Comment:
This will stop the contrail being rendered at all, which surely isn't what we want!?

I assume that we want to be rounding the alpha up and then removing this check.

var alpha = ((renderLength - i) * color.A + renderLength - 1) / renderLength;

Reply:
`wcr.DrawLine(wr.Screen3DPosition(curPos), wr.Screen3DPosition(nextPos), screenWidth, curColor, nextColor);` the previous contrail sequences are rendered.

If the alpha drop to 0, there is no need to render them and, `i` is increasing, if previous `i`  alpha drop to 0, the next `i` will also be 0, thus there is no need to render those which follows.

> var alpha = ((renderLength - i) * color.A + renderLength - 1) / renderLength;
Can be if you want. I will have a test at it.

Reply:
The alpha shouldn't be dropping to 0 until the end of the line, by the construction of the loop.

Reply:
Fixed